### PR TITLE
fix: overly thin font rendering on mojave

### PR DESCRIPTION
**Pull request by codebytere**
##### Description of Change

Resolves https://github.com/electron/electron/issues/14948.

Backports [this CL](https://skia-review.googlesource.com/c/157566/) from skia, which changes behavior to detect if requesting smoothing does not change the rendering, if it changes the rendering, or if it changes the rendering and applies subpixel coverage.

/cc @deepak1556 @MarshallOfSound 

##### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

##### Release Notes

Notes: fix too-thin font rendering in MacOS Mojave

### CODE REVIEW

File: patches/common/skia/fix_font_thickness.patch [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+-    static bool gSupportsLCD = []{
+-        uint32_t bitmap[16][16] = {};
++static SmoothBehavior smooth_behavior() {
++    static SmoothBehavior gSmoothBehavior = []{
++        uint32_t noSmoothBitmap[16][16] = {};
++        uint32_t smoothBitmap[16][16] = {};
+         UniqueCFRef<CGColorSpaceRef> colorspace(CGColorSpaceCreateDeviceRGB());
+-        UniqueCFRef<CGContextRef> cgContext(
+-                CGBitmapContextCreate(&bitmap, 16, 16, 8, 16*4, colorspace.get(), BITMAP_INFO_RGB));
++        SkUniqueCFRef<CGContextRef> noSmoothContext(
```
Comment (deepak1556) [inline patches/common/skia/fix_font_thickness.patch]: `SkUniqueCFRef` was added recently in https://skia-review.googlesource.com/c/skia/+/155612, the skia version on our master doesn't have it, it should be `UniqueCFRef` in this patch.

**Comment by codebytere** [comment]
@cebor i made an attempt this today and it looks like the patch diffs are incompatible between Chromium M62/M66 and M69, so this may be impossible 😢 

I'll try to see what i can do moving forward.

**Comment by dalDevelo** [comment]
Any news on this?

**Comment by dalDevelo** [comment]
@codebytere I'm running the [vscode exploration build](https://github.com/Microsoft/vscode/issues/61787) which is currently using electron 4.0.0-beta 7. However, I'm seeing zero improvement in font rendering on a retina screen compared to the release version. I still have to run `defaults write -g CGFontRenderingFontSmoothingDisabled -bool FALSE` in order for fonts in vscode to display as they did prior to Mojave.

**Comment by Zehua-Chen** [comment]
I tried to load [microsoft.github.io](https://microsoft.github.io/monaco-editor/) using the just-released electron 4 beta 8, on macOS Mojave (10.14.1), but I do not think there are any improvements to font rendering compared to the current vs code release.

# Electron 4 Beta 8
electron 4 beta 8

# VS Code 1.29.1
code

# Safari
safari

**Comment by dalDevelo** [comment]
@Zehua-Chen . I'm guessing the upgrade to Chromiuim 70 #15405 has to happen before we will see any progress on this issue, as @codebytere mentioned above...

> i made an attempt this today and it looks like the patch diffs are incompatible between Chromium M62/M66 and M69, so this may be impossible 😢
>

**Comment by miniak** [comment]
@cebor backported to Electron 3 https://github.com/electron/libchromiumcontent/pull/759

**Comment by miniak** [comment]
maybe we should backport this one as well https://skia-review.googlesource.com/c/skia/+/177880

**Comment by dalDevelo** [comment]
This is not fixed in 4.1.5, as per my comment in #17737.

**Pull request merged**